File: Banker'sAlgo/C/resource_request_algorithm.c
#include <stdio.h>

int isSafe(int n, int m, int need[][m], int alloc[][m], int avail[], int seq[]) {
    int work[m], finish[n], count = 0, i, j;
    for (j = 0; j < m; j++)
        work[j] = avail[j];
    for (i = 0; i < n; i++)
        finish[i] = 0;
    while (count < n) {
        int found = 0;
        for (i = 0; i < n; i++) {
            if (finish[i])
                continue;
            int ok = 1;
            for (j = 0; j < m; j++)
                if (need[i][j] > work[j]) {
                    ok = 0;
                    break;
                }
            if (ok) {
                for (j = 0; j < m; j++)
                    work[j] += alloc[i][j];
                finish[i] = 1;
                seq[count++] = i;
                found = 1;
                break;
            }
        }
        if (!found)
            return 0;
    }
    return 1;
}

int main() {
    int n, m, i, j;
    printf("Enter number of processes: ");
    scanf("%d", &n);
    printf("Enter number of resource types: ");
    scanf("%d", &m);

    int avail[m];
    printf("Enter available instances of each resource: ");
    for (j = 0; j < m; j++)
        scanf("%d", &avail[j]);

    int max[n][m], alloc[n][m], need[n][m];
    printf("Enter max demand matrix (Max):\n");
    for (i = 0; i < n; i++)
        for (j = 0; j < m; j++)
            scanf("%d", &max[i][j]);
    printf("Enter allocation matrix (Allocation):\n");
    for (i = 0; i < n; i++)
        for (j = 0; j < m; j++)
            scanf("%d", &alloc[i][j]);
    for (i = 0; i < n; i++)
        for (j = 0; j < m; j++)
            need[i][j] = max[i][j] - alloc[i][j];

    int p;
    printf("Enter process making request (1-%d): ", n);
    scanf("%d", &p);
    p--;
    int req[m];
    printf("Enter request vector: ");
    for (j = 0; j < m; j++)
        scanf("%d", &req[j]);

    for (j = 0; j < m; j++) {
        if (req[j] > need[p][j]) {
            printf("ERROR: request exceeds maximum claim of P%d. Process must be terminated.\n", p + 1);
            return 0;
        }
    }

    for (j = 0; j < m; j++) {
        if (req[j] > avail[j]) {
            printf("Request of P%d EXCEEDS available resources. Process must wait.\n", p + 1);
            return 0;
        }
    }

    int savedAvail[m];
    int savedAlloc[n][m], savedNeed[n][m];
    for (j = 0; j < m; j++)
        savedAvail[j] = avail[j];
    for (i = 0; i < n; i++)
        for (j = 0; j < m; j++) {
            savedAlloc[i][j] = alloc[i][j];
            savedNeed[i][j] = need[i][j];
        }

    for (j = 0; j < m; j++) {
        avail[j] -= req[j];
        alloc[p][j] += req[j];
        need[p][j] -= req[j];
    }

    int seq[n];
    int safe = isSafe(n, m, need, alloc, avail, seq);

    if (safe) {
        printf("Request of P%d is GRANTED. System remains in SAFE state.\n", p + 1);
        printf("Safe sequence: ");
        for (i = 0; i < n; i++)
            printf("P%d%s", seq[i] + 1, i < n - 1 ? " -> " : "\n");
        printf("New Available:\t");
        for (j = 0; j < m; j++)
            printf("%d\t", avail[j]);
        printf("\n");
    } else {
        for (j = 0; j < m; j++) {
            avail[j] = savedAvail[j];
            alloc[p][j] = savedAlloc[p][j];
            need[p][j] = savedNeed[p][j];
        }
        printf("Request of P%d is DENIED. Granting it would leave the system UNSAFE.\n", p + 1);
        printf("State rolled back; process must wait.\n");
    }
    return 0;
}

File: Banker'sAlgo/ResourceRequestAlgorithm.java
import java.util.Arrays;
import java.util.Scanner;

public class ResourceRequestAlgorithm {

    static boolean isSafe(int n, int m, int[][] need, int[][] alloc, int[] avail, int[] seq) {
        int[] work = new int[m];
        boolean[] finish = new boolean[n];
        int count = 0;
        for (int j = 0; j < m; j++)
            work[j] = avail[j];
        while (count < n) {
            boolean found = false;
            for (int i = 0; i < n; i++) {
                if (finish[i]) continue;
                boolean ok = true;
                for (int j = 0; j < m; j++)
                    if (need[i][j] > work[j]) { ok = false; break; }
                if (ok) {
                    for (int j = 0; j < m; j++)
                        work[j] += alloc[i][j];
                    finish[i] = true;
                    seq[count++] = i;
                    found = true;
                    break;
                }
            }
            if (!found) return false;
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter number of processes: ");
        int n = sc.nextInt();
        System.out.print("Enter number of resource types: ");
        int m = sc.nextInt();

        int[] avail = new int[m];
        System.out.print("Enter available instances of each resource: ");
        for (int j = 0; j < m; j++)
            avail[j] = sc.nextInt();

        int[][] max = new int[n][m];
        System.out.println("Enter max demand matrix (Max):");
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                max[i][j] = sc.nextInt();

        int[][] alloc = new int[n][m];
        System.out.println("Enter allocation matrix (Allocation):");
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                alloc[i][j] = sc.nextInt();

        int[][] need = new int[n][m];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                need[i][j] = max[i][j] - alloc[i][j];

        System.out.print("\nEnter process making request (1-" + n + "): ");
        int p = sc.nextInt() - 1;
        System.out.print("Enter request vector: ");
        int[] req = new int[m];
        for (int j = 0; j < m; j++)
            req[j] = sc.nextInt();

        boolean exceedsNeed = false;
        for (int j = 0; j < m; j++)
            if (req[j] > need[p][j]) { exceedsNeed = true; break; }

        if (exceedsNeed) {
            System.out.println("ERROR: request exceeds maximum claim of P" + (p + 1) + ". Process must be terminated.");
            sc.close();
            return;
        }

        boolean mustWait = false;
        for (int j = 0; j < m; j++)
            if (req[j] > avail[j]) { mustWait = true; break; }

        if (mustWait) {
            System.out.println("Request of P" + (p + 1) + " EXCEEDS available resources. Process must wait.");
            sc.close();
            return;
        }

        int[] savedAvail = Arrays.copyOf(avail, m);
        int[][] savedAlloc = new int[n][m];
        int[][] savedNeed = new int[n][m];
        for (int i = 0; i < n; i++) {
            savedAlloc[i] = Arrays.copyOf(alloc[i], m);
            savedNeed[i] = Arrays.copyOf(need[i], m);
        }

        for (int j = 0; j < m; j++) {
            avail[j] -= req[j];
            alloc[p][j] += req[j];
            need[p][j] -= req[j];
        }

        int[] seq = new int[n];
        boolean safe = isSafe(n, m, need, alloc, avail, seq);

        if (safe) {
            System.out.println("Request of P" + (p + 1) + " is GRANTED. System remains in SAFE state.");
            System.out.print("Safe sequence: ");
            for (int k = 0; k < n; k++)
                System.out.print("P" + (seq[k] + 1) + (k < n - 1 ? " -> " : "\n"));
            System.out.print("\nNew Available:\t");
            for (int j = 0; j < m; j++)
                System.out.print(avail[j] + "\t");
            System.out.println();
        } else {
            for (int j = 0; j < m; j++) {
                avail[j] = savedAvail[j];
                alloc[p][j] = savedAlloc[p][j];
                need[p][j] = savedNeed[p][j];
            }
            System.out.println("Request of P" + (p + 1) + " is DENIED. Granting it would leave the system UNSAFE.\nState rolled back; process must wait.");
        }
        sc.close();
    }
}